python/dask_cudf/dask_cudf/tests/test_groupby.py

```diff
@@ -27,92 +41,75 @@ def pdf():
             "y": np.random.normal(size=10000),
         }
     )
-    return pdf
 
+    # insert nulls into dataframe at random
+    if request.param == "null":
+        pdf = pdf.mask(np.random.choice([True, False], size=pdf.shape))
+
+    return pdf
 
-@pytest.mark.parametrize("aggregation", AGGS)
-@pytest.mark.parametrize("series", [False, True])
-def test_groupby_basic(series, aggregation, pdf):
 
+@pytest.mark.parametrize("aggregation", ALL_AGGS)
+@pytest.mark.parametrize("type", ["frame", "series"])
+def test_groupby_basic(type, aggregation, pdf):
     gdf = cudf.DataFrame.from_pandas(pdf)
     gdf_grouped = gdf.groupby("xx")
     ddf_grouped = dask_cudf.from_cudf(gdf, npartitions=5).groupby("xx")
 
-    if series:
+    if type == "series":
         gdf_grouped = gdf_grouped.xx
         ddf_grouped = ddf_grouped.xx
 
-    a = getattr(gdf_grouped, aggregation)()
-    b = getattr(ddf_grouped, aggregation)().compute()
+    check_dtype = aggregation != "count"
 
-    if aggregation == "count":
-        dd.assert_eq(a, b, check_dtype=False)
-    else:
-        dd.assert_eq(a, b)
+    expect = getattr(gdf_grouped, aggregation)()
+    actual = getattr(ddf_grouped, aggregation)()
 
-    a = gdf_grouped.agg({"xx": aggregation})
-    b = ddf_grouped.agg({"xx": aggregation}).compute()
+    assert_cudf_groupby_layers(actual)
 
-    if aggregation == "count":
-        dd.assert_eq(a, b, check_dtype=False)
+    if aggregation == "cumsum" and type == "series":
+        with pytest.xfail(reason="https://github.com/dask/dask/issues/9313"):
+            dd.assert_eq(expect, actual, check_dtype=check_dtype)
     else:
-        dd.assert_eq(a, b)
+        dd.assert_eq(expect, actual, check_dtype=check_dtype)
 
+    expect = gdf_grouped.agg({"xx": aggregation})
+    actual = ddf_grouped.agg({"xx": aggregation})
 
-@pytest.mark.parametrize("series", [True, False])
-@pytest.mark.parametrize("aggregation", CUMULATIVE_AGGS)
-def test_groupby_cumulative(aggregation, pdf, series):
```

Comment:
Merged this test into `test_groupby_basic`, as it seems like this is doing roughly the same thing but just with `CUMULATIVE_AGGS`; also added `CUMULATIVE_AGGS` to the set of aggregations being tested in `test_groupby_agg`.

While doing this, seemed to uncover some issues some issues with the cumulative aggregations:

- using `cumsum()` or `cumcount()` was using Dask's groupby implementation instead of dask-cudf
- after adding the appropriate methods to dask-cudf's groupby/series objects, we still seem to be getting some assertion errors

cc @brandon-b-miller since you initially added the set of cumulative aggregations to dask-cudf

Reply:
I think that based off the docstring for `groupby_agg` that the algorithm wouldn't necessarily support cumulative operations, hence falling back to the upstream implementation. The failed assertions that I assume you are referring to seemed to be evidence of that at the time. Indeed for cumulative operations I'd expect generalizations to be needed somewhere to account for extra information the partitions need to share between them. Not sure if `groupby_agg` is set up to do this.

Reply:
In that case, would it make sense to remove the cumulative aggregations from `SUPPORTED_AGGS` so that we always use the upstream codepath for them? I had assumed that #10889 was adding support for _dask-cudf optimized_ cumulative aggregations, but looks like instead we just wanted to verify that cumulative aggregations worked on a dask-cudf groupby object in general?

Reply:
My view on it was while generalizing `groupby_agg` for this would be optimal, it was still worth opening up the pipes for dask to do it's normal approach with `cudf` partitions on the back end. It could maybe have done with a follow up issue admittedly.

Reply:
Yeah I agree that having the tests in place to ensure that the aggregations at least work in upstream Dask with cuDF backend makes sense (might also be worth adding tests like this for other non-optimized aggregations). My issue is that based on the test failures here, it seems like attempting to do an `agg()` operation with `"cumsum"` or `"cumcount"` will give the incorrect result since they're included in `SUPPORTED_AGGS` and wind up going through `groupby_agg` where they're not supported.

How would you feel about keeping the cumulative aggregation tests separate (maybe linking to an issue around the follow-up work), adding in some basic tests for `agg()` using the cumulative aggregations, and removing the cumulative aggs from the list of `SUPPORTED_AGGS` so that the agg tests pass?

Might also be worth renaming functions / objects around "supported" aggs in dask-cudf to _optimized_ aggs so the distinction between the two is more clear.

Reply:
I'm pretty onboard with that. I really like `OPTIMIZED_AGGS` as well 👍